agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/utils/SanitizationHelper.java

```diff
@@ -0,0 +1,143 @@
+    }
+    Map<String, String> sanitized = new HashMap<>();
+    for (Map.Entry<String, String> entry : properties.entrySet()) {
+      String sanitizedKey = sanitizeKey(entry.getKey());
+      String sanitizedValue = sanitizeValue(entry.getValue());
+      if (!Strings.isNullOrEmpty(sanitizedValue)) {
+        String uniqueKey = makeKeyUnique(sanitizedKey, sanitized);
+        sanitized.put(uniqueKey, sanitizedValue);
+      }
+    }
+    return sanitized;
+  }
+
+  private static boolean needsSanitizing(Map<String, String> properties) {
+    for (Map.Entry<String, String> entry : properties.entrySet()) {
+      String key = entry.getKey();
+      String value = entry.getValue();
+      if (Strings.isNullOrEmpty(key)
+          || Strings.isNullOrEmpty(value)
+          || key.length() > MAX_KEY_LENGTH
+          || value.length() > MAX_VALUE_LENGTH) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /** Function to create unique key. */
+  private static String makeKeyUnique(String key, Map<String, ?> map) {
+    if (!map.containsKey(key)) {
+      return key;
+    }
+    String truncatedKey = Strings.truncate(key, MAX_KEY_LENGTH - 3);
+    int candidate = 1;
+    do {
+      key = truncatedKey + candidate;
+      ++candidate;
+    } while (map.containsKey(key));
+    return key;
+  }
+
+  /** Function to sanitize value. */
+  private static String sanitizeValue(String value) {
+    return Strings.trimAndTruncate(value, MAX_VALUE_LENGTH);
+  }
+
+  /** Function to sanitize key. */
+  private static String sanitizeKey(String key) {
+    String sanitizedKey = Strings.trimAndTruncate(key, MAX_KEY_LENGTH);
+    return makeKeyNonEmpty(sanitizedKey);
+  }
+
+  /** Function to return non empty key. */
+  private static String makeKeyNonEmpty(String sanitizedKey) {
+    // TODO check if this is valid. Below code is based on .Net implementation of the same
+    return Strings.isNullOrEmpty(sanitizedKey) ? "empty" : sanitizedKey;
+  }
+
+  /** Function to sanitize key value pair in measurements. */
+  public static void sanitizeMeasurements(Map<String, Double> measurements) {
+    if (measurements == null) {
+      return;
+    }
+    Map<String, SanitizedEntry<Double>> sanitizedMeasurements = new HashMap<>();
+    for (Map.Entry<String, Double> entry : measurements.entrySet()) {
+      String sanitizedKey = sanitizeKey(entry.getKey());
+      // TODO sanitize value ?
+      if (!sanitizedKey.equals(entry.getKey())) {
+        sanitizedMeasurements.put(
+            entry.getKey(), new SanitizedEntry<>(sanitizedKey, entry.getValue()));
+      }
+    }
+    for (Map.Entry<String, SanitizedEntry<Double>> entry : sanitizedMeasurements.entrySet()) {
+      measurements.remove(entry.getKey());
+      String uniqueKey = makeKeyUnique(entry.getValue().getKey(), measurements);
+      measurements.put(uniqueKey, entry.getValue().getValue());
+    }
+  }
+
+  private SanitizationHelper() {}
+
+  static class SanitizedEntry<V> {
```

Comment:
this is only applied to measurements, not other entry... SanitizedPair?

Reply:
removed this completely in the next commit